### PR TITLE
Node.js / Webassembly Javascript Promise Integration support

### PR DESCRIPTION
**Pull request by joemarshall**
This PR adds support to the emscripten/pyodide backend for Javascript Promise Integration, a new feature in webassembly which allows webassembly code (i.e. python) to call asynchronous javascript apis like fetch. This means that on browsers where it is supported, urllib3 works in a much simpler way, directly with the browser fetch api. 

As a bonus, this means it is possible to use urllib3 in node.js, albeit with an experimental command line flag on the current stable version of node.

This is what I was talking about in #3400

### CODE REVIEW

**Comment by sethmlarson** [comment]
@joemarshall Thank you for opening this PR, this is really exciting stuff! I'll give this a full review later. I think my high-level comment for now is to add more comments inline in the source code to explain what's happening and why. This will help us immensely for fixing issues in the future :pray:

**Comment by joemarshall** [comment]
@sethmlarson I added tests + more commenting. The tests won't work within nox until I merge https://github.com/pyodide/pytest-pyodide/pull/140 but it is probably worth a code review at this point so that it can be mergable at the point that pull hits pytest-pyodide.

**Comment by joemarshall** [comment]
@sethmlarson I pushed the change into pytest-pyodide and finished the ci here and it now hits 100% coverage and emscripten tests pass on node, chrome and firefox. It failed on something unrelated in Ubuntu this last run which I think might be a CI weirdness, but overall I think is ready for review now.

**Review by illia-v — COMMENTED** [review]
@joemarshall thank you for your work on this! I'm looking forward to having this feature integrated.

I noticed that the PR includes adjustments due to urllib3 being built in Pyodide and significant changes to the testing code. Would it be feasible to split the PR into a few smaller ones to facilitate a thorough review and faster integration?

Perhaps one PR could focus on changing the documentation and tests to use the built-in urllib3, another to replace webdriver-manager with Selenium, another on installing Node.js and testing non-JSPI code failing in Node.js, and a final one for the JSPI code itself.

**Comment by sethmlarson** [comment]
Thanks for the review @illia-v! @joemarshall are you still interested in landing this PR?

File: docs/reference/contrib/emscripten.rst
Comment (hoodmane) [reply docs/reference/contrib/emscripten.rst:89]: Would also be good to use the standard capitalization of "WebAssembly".

Comment (hoodmane) [reply docs/reference/contrib/emscripten.rst:89]: Maybe mention here also that it requires Node >= 20.

File: dummyserver/app.py
Comment (hoodmane) [inline dummyserver/app.py:384]: What is rewriting the lockfile needed for? In `emscripten/conftest.py`, you run `loadPackage` the `urlib3.whl` directly, which I'd think would be good enough to make sure you're getting the right version. Is there some place that installs it by name?

File: src/urllib3/contrib/emscripten/fetch.py
Comment (hoodmane) [inline src/urllib3/contrib/emscripten/fetch.py:669]: I think it would make sense to just drop this polyfill now and only support JSPI on Pyodide >= 0.26.2.
```suggestion
        from pyodide.ffi import can_run_sync
```

File: src/urllib3/contrib/emscripten/fetch.py
Comment (hoodmane) [inline src/urllib3/contrib/emscripten/fetch.py:672]: I'd think we would only want to catch `Exception` here, wouldn't want to catch a `KeyboardInterrupt` or `SystemExit`? Or maybe even just catch `ImportError`?
```suggestion
    except Exception:
```

File: test/contrib/emscripten/conftest.py
Comment (hoodmane) [inline test/contrib/emscripten/conftest.py:165]:
```suggestion
        if not has_jspi:
```

File: test/contrib/emscripten/conftest.py
Comment (hoodmane) [inline test/contrib/emscripten/conftest.py:166]: Why are we monkey patching `has_jspi()` here? Shouldn't `has_jspi()` correctly detect this? In real life if people are using Firefox or Safari or some other runtime without JSPI, they won't get this monkey patch.

File: src/urllib3/contrib/emscripten/fetch.py
Comment (hoodmane) [inline src/urllib3/contrib/emscripten/fetch.py:441]: We may want to mention here that node >= 20 is required. We're not sure that node 24 won't require the flag yet...

File: src/urllib3/contrib/emscripten/fetch.py
Comment (hoodmane) [inline src/urllib3/contrib/emscripten/fetch.py:643]: Nice!

File: src/urllib3/contrib/emscripten/fetch.py
Comment (hoodmane) [inline src/urllib3/contrib/emscripten/fetch.py:599]: The `arrayBuffer()` call isn't cancelable so I think we want to just use `run_sync` directly here.
```suggestion
        body = run_sync(response_js.arrayBuffer()).to_py()
```

File: src/urllib3/contrib/emscripten/fetch.py
Comment (hoodmane) [inline src/urllib3/contrib/emscripten/fetch.py:357]: We should probably call `self.js_read_stream.cancel()` when we're done with it. Also a nit: it reads a bit nicer with an early return.
```suggestion
    def close(self) -> None:
        if self.is_closed():
            return
        self._is_closed = True
        self._is_done = True
        self.read_len = 0
        self.read_pos = 0
        self.js_read_stream.cancel()
        self.js_read_stream = None
        self.request = None
        self.response = None
        super().close()
```

File: src/urllib3/contrib/emscripten/fetch.py
Comment (joemarshall) [reply src/urllib3/contrib/emscripten/fetch.py:615]: https://developer.mozilla.org/en-US/docs/Web/API/Response/arrayBuffer says it can return AbortError, so I figure at least mozilla think that it can be cancelled between header and response fetch, although I haven't delved into browser code.

File: src/urllib3/contrib/emscripten/fetch.py
Comment (joemarshall) [reply src/urllib3/contrib/emscripten/fetch.py:451]: done

File: test/contrib/emscripten/conftest.py
Comment (joemarshall) [reply test/contrib/emscripten/conftest.py:174]: It was just to save loading different browsers and force the jspi code to be executed. I simplified the code relating to this now, so it uses the non-jspi selenium instead.

**Comment by joemarshall** [comment]
@sethmlarson @illia-v I fixed various issues from the reviews above and simplified this change somewhat. I'm not 100% sure how/whether to split it up into multiple PRs because everything is slightly entangled. Any comments would be welcome.

File: .github/workflows/ci.yml
Comment (hoodmane) [inline .github/workflows/ci.yml:114]: What version of chrome was in there before the version override? Presumably an old one? We might want to check if it works against chrome 126 since there are some jspi changes there.

File: emscripten-requirements.txt
Comment (hoodmane) [inline emscripten-requirements.txt:2]: With the pyodide-build unvendoring, do we need a pin here @ryanking13? Should it pin latest?

File: src/urllib3/contrib/emscripten/fetch.py
Comment (hoodmane) [reply src/urllib3/contrib/emscripten/fetch.py:615]: Cool. Indeed if you poke around the fetch spec long enough you can find that the spec says this should happen:
> To abort a fetch() call with a promise, request, responseObject, and an error:
> ...
> 5. If response’s [body](https://fetch.spec.whatwg.org/#concept-response-body) is non-null and is [readable](https://streams.spec.whatwg.org/#readablestream-readable), then [error](https://streams.spec.whatwg.org/#readablestream-error) response’s [body](https://fetch.spec.whatwg.org/#concept-response-body) with error.


https://fetch.spec.whatwg.org/#abort-fetch

File: test/contrib/emscripten/conftest.py
Comment (hoodmane) [inline test/contrib/emscripten/conftest.py:145]: It might be easier to replace `webworker.js`. It's only 26 lines and quite simple. But this seems fine if a bit regrettable.

File: test/contrib/emscripten/test_emscripten.py
Comment (hoodmane) [inline test/contrib/emscripten/test_emscripten.py:30]: Is this intended to support multiple versions of `pytest-pyodide`? Seems like you might as well pin the new version and drop the old code.

File: test/contrib/emscripten/test_emscripten.py
Comment (hoodmane) [inline test/contrib/emscripten/test_emscripten.py:692]: Unrelated to this PR, but it should work just fine to do:
```py
@run_in_pyodide
async def test_something(...):
    await some_async_call()
```
See for instace: https://github.com/pyodide/pyodide/blob/main/src/tests/test_browser_apis.py

**Review by hoodmane — APPROVED** [review]
Looks good to me. I made some more very minor comments but none of them need to be addressed necessarily in my opinion.

On code size, it is true that it's almost a 1000 line diff, but I don't find it too hard to review. 80% of the changes are to `contrib/emscripten` and the ones that aren't are to ci, noxfile, dummyserver, docs stuff that doesn't go at runtime. So I don't personally see a good reason to split it up, but of course we'll have to see what the maintainers want.

Thanks @joemarshall!

File: test/contrib/emscripten/conftest.py
Comment (joemarshall) [reply test/contrib/emscripten/conftest.py:253]: There is a way to disable it but it is deep in the firefox profile meaning it would require code in pytest-pyodide or monkeypatching; making it be on the same origin is easier.

File: emscripten-requirements.txt
Comment (ryanking13) [reply emscripten-requirements.txt:2]: You don't need to set the pyodide-build version equal to the target Pyodide version anymore, but I think it is not a bad thing to pin the version for reproducible build.

Comment (joemarshall) [reply emscripten-requirements.txt]: Actually, thinking about it, I don't need pyodide-build any more, not sure that should be in really. Will fix.

File: test/contrib/emscripten/conftest.py
Comment (joemarshall) [reply test/contrib/emscripten/conftest.py]: yeah, I did that now.

File: test/contrib/emscripten/test_emscripten.py
Comment (joemarshall) [reply test/contrib/emscripten/test_emscripten.py:689]: yeah, I just didn't want to have to add an asyncio test plugin just for this single test

File: .github/workflows/ci.yml
Comment (joemarshall) [reply .github/workflows/ci.yml:114]: It will only work against 126 once pyodide 0.27 is properly released, hence I stuck it back on 120 for testing for now, because getting the latest version of chrome breaks everything. I figure it doesn't matter, because the next version of pyodide that includes an upgraded urllib3 will also have the fix for jspi in pyodide (the feature detect thing)

**Comment by hoodmane** [comment]
@sethmlarson would you mind reviewing this PR when you have a chance? Or is there another maintainer that could do the review?

File: .github/workflows/ci.yml
Comment (illia-v) [inline .github/workflows/ci.yml:136]: I believe the key has to include a hash of `noxfile.py` which contains a Pyodide version.
If the key never changes, the following can happen: when Pyodide version is changed, files of the new version won't be added to the cache until the cache is evicted. And the eviction may never happen if the cache is restored frequently enough.

```suggestion
          # noxfile.py contains a Pyodide version used currently.
          key: pyodide-downloads-${{ hashFiles('noxfile.py') }}
```

File: docs/reference/contrib/emscripten.rst
Comment (illia-v) [inline docs/reference/contrib/emscripten.rst:31]:
```suggestion
One minor note - when running Pyodide code from JavaScript, if you use ``pyodide.runPythonAsync`` rather
```

File: src/urllib3/contrib/emscripten/fetch.py
Comment (illia-v) [inline src/urllib3/contrib/emscripten/fetch.py:669]: Is this a TODO?

**Review by illia-v — COMMENTED** [review]
@hoodmane thank you for you review!

File: .github/workflows/ci.yml
Comment (illia-v) [reply .github/workflows/ci.yml:136]: Also, this step should be limited to the `emscripten` session

File: .github/workflows/ci.yml
Comment (joemarshall) [reply .github/workflows/ci.yml]: fixed with 0.26.3 I think

**Comment by joemarshall** [comment]
@illia-v Thanks for the review, I think I've addressed all those comments. CI passes okay.

**Review by illia-v — APPROVED** [review]
@joemarshall thanks again!

FYI, Firefox and Chrome tests have been split in separate jobs in the main branch to get faster CI (see #3505), so I made the Node.js job separate while I was resolving conflicts in the last merge commit

**Pull request merged**
